lib/xgu/xgu.h

```diff
@@ -0,0 +1,535 @@
+    return push_command_parameter(p, NV097_SET_STENCIL_OP_zfail, so);
+}
+
+inline
+DWORD* xgu_set_stencil_op_zpass(DWORD* p, XguStencilOp so) {
+    return push_command_parameter(p, NV097_SET_STENCIL_OP_zpass, so);
+}
+
+/* ==== Vertex Data Array ==== */
+
+inline
+DWORD* xgu_set_vertex_data_array_format(DWORD* p, XguVertexArray index, XguVertexArrayType format,
+                                        uint32_t size, uint32_t stride) {
+    return push_command_parameter(p, NV097_SET_VERTEX_DATA_ARRAY_FORMAT + index*4,
+                                  XGU_MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE, format) |
+                                  XGU_MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_SIZE, size) |
+                                  XGU_MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_STRIDE, stride));
+}
+
+inline
+DWORD* xgu_set_vertex_data_array_offset(DWORD* p, XguVertexArray index, const uint32_t data) {
+    return push_command_parameter(p, NV097_SET_VERTEX_DATA_ARRAY_OFFSET + index*4,
+                                  data/* & 0x03fffff*/);
+}
+
+inline
+DWORD* xgu_elements16(DWORD* p, uint16_t* elements, uint32_t count) {
+    assert(count % 2 == 0);
+    // This one is a bit complicated
+    assert(false);
+    return p;
+}
+
+inline
+DWORD* xgu_elements32(DWORD* p, uint32_t* elements, uint32_t count) {
+    // This one is a bit complicated
+    assert(false);
+    return p;
+}
+
+inline
+DWORD* xgu_draw_arrays(DWORD* p, uint32_t start, uint32_t count) {
+    return push_command_parameter(p, 0x40000000|NV097_DRAW_ARRAYS,
+                                  XGU_MASK(NV097_DRAW_ARRAYS_COUNT, (count-1)) |
+                                  XGU_MASK(NV097_DRAW_ARRAYS_START_INDEX, start));
+}
+
+/* ==== Alpha/Blend/Cull ==== */
+inline
+DWORD* xgu_set_alpha_test_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_ALPHA_TEST_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_blend_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_BLEND_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_cull_face_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_CULL_FACE_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_depth_test_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_DEPTH_TEST_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_dither_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_DITHER_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_lighting_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_LIGHTING_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_stencil_test_enable(DWORD* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_STENCIL_TEST_ENABLE, enabled);
+}
+
+inline
+DWORD* xgu_set_alpha_func(DWORD* p, uint32_t func) {
+    return push_command_parameter(p, NV097_SET_ALPHA_FUNC, func);
+}
+
+inline
+DWORD* xgu_set_alpha_ref(DWORD* p, uint32_t ref) {
```

Comment:
I don't think this is a full 32 bit value; probably a float or `uint8_t` or something; check how XQEMU handles this